### DISCUSSION TITLE
Guidance on setting up a beginner-user environment?

### DISCUSSION THREAD

**Discussion by aalexandersson** [moj-analytical-services/splink#2306] [Q&A] [answered]
I'm currently trying to set up an environment for a couple of beginner users to work on Splink 4 for the first time, and I am struggling a bit; for context, I have about 2 years of experience with Splink 3 myself. In contrast, for guidance on setting up a development environment, see discussion [#2019](#2019 ). The beginner-user setup will be in one or two Windows 10 desktop virtual machines (VM).  

Here are some of the choice considerations I hope to get feedback on:

1. One VM with Python 3.11 for actual use (production) and one VM with Python 3.12 for "sandbox" (testing/development)? With Python 3.13 coming out in October, would you instead recommend only Python 3.12 for both VMs? Or something else?
2. How to update Python patches easily and safely, for example, from 3.12.4 to 3.12.5?
3. Would you recommend using virtual environments with [venv](https://docs.python.org/3/library/venv.html) for beginner user projects? Using `venv` can be helpful but it does add complexity; I want to keep it simple for beginner users.
4. Would you manage required packages in `requirements.txt`, for example as instructed by Microsoft for Visual Studio [here](https://learn.microsoft.com/en-us/visualstudio/python/managing-required-packages-with-requirements-txt)? I use `requirements.txt` myself but, like `venv`, it adds complexity.
5. How important is the choice of IDE? My current IDE is [RStudio](https://docs.posit.co/ide/user/ide/get-started/). I also consider [Visual Studio Code](https://code.visualstudio.com/) (VS Code), [Theia IDE](https://theia-ide.org/) and [Positron IDE](https://github.com/posit-dev/positron).
6. Other than the above issues and the Splink documentation, is there anything else that you think would be very helpful for a new user? I am familiar with [Quarto](https://quarto.org/) for automated reporting and will probably set it up. I also have LaTeX setup for PDF output but I am concerned it will cause more trouble than worth for beginner users.

Aside: I just installed Splink 4, and I am eager to try it out. Thanks Robin and everyone else for your hard work!

**Comment by ADBond** [comment, marked as answer]
Just my ha'pennyworth, but I’ll try to address your queries. I’m not quite sure whether you mean setting things up working on a data-linking project that uses (installed) Splink, or for working on contributing for Splink itself, but I’ll answer primarily assuming the former, but also give info for the latter.

1. I think it would probably best to stick to the same version of python throughout, so that your environment is consistent. Then you are less likely to have e.g. code that works in one environment but not the other, for instance if it uses newer features. Personally I would use the latest version that works with any other constraints you may have - one thing to note is that I don’t think python 3.12 is currently working with `pyspark` (or at least not straightforwardly). Generally each minor release will have some useful new features that it is handy to have access to. For Splink dev the version doesn’t matter so much, as we support all in-life python versions, so can’t use any features that are not available in python 3.8 (currently). 
2. Afraid I personally haven’t really found a clean workflow for this. When I do so I generally uninstall/reinstall
3. Yep, absolutely. It does add a little complexity, but I feel like it is just a small additional complexity amongst many when getting started with something new, and things can get complicated/confusing/messy if things go wrong without a virtual environment. Plus it is good to start with good habits.
4. Again, I think this (or some other equivalent solution) is pretty essential. It’s very easy for people’s environments to get subtly out of sync, so having a crystal-clear set of installed packages (and that can be used for testing etc) means everyone is on the same page. In Splink we use [poetry](https://python-poetry.org/) , but for a straightforward project I would just stick with a simple `requirements.txt`. If things go in complexity you can always revise how you manage dependencies. With this it's super simple, if your environment ends up getting weird for some reason, to just delete the old virtual environment, make a fresh one, and get all the dependencies back to where they should be.
5. This probably comes down a bit more to personal preference. I don’t think it matters hugely, as long as the IDE comes with tools that help with developer workflows (linters, integration with testing/debugging/source control - again, the kinds of things people find helpful vary from person to person). It would probably be helpful in terms of troubleshooting and the like for people to use the same one, but not necessarily essential, and so may depend on what people are already familiar with, if any. FWIW most, if not all, of our team uses VSCode in the main.
6. Can’t think of anything else particularly helpful for beginners. More generally for starting something new, a linter is useful for ensuring code style consistency, and an autoformatter also. We use [ruff](https://docs.astral.sh/ruff/) for both. I also think that including type hints (and [mypy](https://mypy.readthedocs.io/en/stable/)) from an early stage helps code readability, and can help beginners think more clearly about what they are trying to do. And doing everything in git from the start 🙂 

In general I would say that it would be helpful to have a clear list of step-by-step instructions for getting set up that explain _what_ to do (and briefly what that step is doing, e.g. 'Set up a virtual environment by running `python3 -m venv venv`'), but putting more detail / info on _why_ to do these steps elsewhere. Usually that info is a lot to take in, and may make more sense to slowly digest once people are up-and-running with the setup, so having an instruction list that can be blindly followed without requiring too much thought/understanding can be a bit less overwhelming

**Reply by aalexandersson** [reply]
Thank you so much, very helpful! I mean setting things up for beginners working on a data linkage project that uses installed Splink.

1. We will use the default DuckDB backend. I will test Python 3.12 a bit. If I do not find a major issue, then I will stick to the new 3.12.
2. Stack Overflow question [45137395](https://stackoverflow.com/questions/45137395/how-do-i-upgrade-the-python-installation-in-windows-10) elaborates on the grade differences between minor releases and patch releases. A new install is sufficient for updating a patch release: from 3.x.y to 3.x.z. 
3. OK, I will use `venv` for this. The documentation at [https://docs.python.org/3/library/venv.html](https://docs.python.org/3/library/venv.html) seems less prescriptive, so I was not sure.
4. OK, I will use `requirements.txt` for this, 
5. VS Code is the most popular IDE amongst Python [Stackoverflow users](https://survey.stackoverflow.co/2024/technology#1-integrated-development-environment).
6. At some point in the future, I hope to develop a [devcontainer](https://containers.dev/) in [Bluefin DX](https://docs.projectbluefin.io/bluefin-dx), to simplify the setup for beginner users; the only user requirement is an IDE that supports devcontainers such as VS Code. We have nothing in Git yet, but I will consider creating a [private repository](https://docs.github.com/en/repositories/creating-and-managing-repositories/about-repositories) in Git for our Splink use template.

**Reply by ADBond** [reply]
Great, that all sounds good. Feel free to reach out again if you have any other queries!
